### PR TITLE
[OSDEV-2784] Add /ship skill and command — full ticket pipeline

### PR DESCRIPTION
**Pull request by TylerHeath1**
## Summary

Adds the `/ship` skill at `.agent/skills/ship/SKILL.md` and a thin command wrapper at `.claude/commands/ship.md`. The skill body is the canonical source of truth (works in any harness — Cursor, Claude Code, etc.); the command is a Claude Code convenience trigger.

`/ship` is the full OS Hub ticket pipeline — reads a Jira ticket, plans, implements, tests, previews, and opens a PR. It orchestrates the existing skills (`oshub-codebase-expert`, `pr-description`, `release-notes`) rather than restating their conventions.

## What's in the skill

**Three entry points** (Step 0):
- ` /ship OSDEV-####` — standard path
- `/ship` with cached plan from `/plan` — skips planning and jumps to building
- `/ship <freeform idea>` — routes to `jira-ticket-writer` (or asks user to create the ticket manually until that skill exists)

**Experience-level calibration** (Step 1):
- Level 1 (never coded): gloss every git/GitHub term at first appearance, slow pacing
- Level 2 (some comfort): gloss only non-obvious terms (PR, merge)
- Level 3 (engineer): no glosses, no narration of routine commands

**8 user checkpoints:**
1. Cached plan detected — use or re-plan?
2. Plan approval — show full plan, wait for 'go'
3. Files review — before tests run
4. Tests result — commit or fix
5. Push branch — explicit yes/no
6. Local preview — verify the change works locally (or honestly report if it can't)
7. Create PR — explicit yes/no
8. Handoff

**Skills it orchestrates:**
- `oshub-codebase-expert` sub-agent for codebase research
- `release-notes` skill for the RELEASE-NOTES.md entry
- `pr-description` skill for the PR body
- An orchestrator-context comment posted on the PR with plan source, Codebase Expert report, and identified risks — gives reviewers the full trail of what /ship did

## Design principles baked in

- **Surface assumptions before coding** — if the ticket is ambiguous, ask before planning
- **Minimum work to satisfy the ticket** — no speculative refactors, no formatting changes outside the plan
- **Surgical changes for implementation agents** — they're explicitly told to touch only what the plan calls for
- **Verifiable check at every step** — tests must pass with zero failures, preview must actually verify, etc.

## How to test

Cannot test the full flow here (would need a real ticket and a full local environment). The structural validation:
- [ ] YAML frontmatter parses correctly in both files
- [ ] `name` and `description` fields are present in skill YAML
- [ ] `argument-hint` is present in command YAML (Claude Code convention)
- [ ] Skill body references existing skills by correct path
- [ ] Command wrapper relative path resolves to the skill file

## Follow-ups (not in this PR)

- Add `memory: project` to the `oshub-codebase-expert` sub-agent so it accumulates OS Hub-specific knowledge across `/ship` runs (separate small PR once #1013 merges)
- Worktree isolation for parallel implementation agents (Step 7) — eliminates file conflict risk when backend + frontend agents overlap
- `PostToolUse` lint hook for implementation agents — catches errors during implementation instead of at Step 9
- Build `test-writer` and `jira-ticket-writer` skills, then update Step 9 and Step 0 to call into them

## Part of

The Common Dev Brain initiative — Epic [OSDEV-2719](https://opensupplyhub.atlassian.net/browse/OSDEV-2719).

- Jira story: [OSDEV-2784](https://opensupplyhub.atlassian.net/browse/OSDEV-2784)
- Confluence: https://opensupplyhub.atlassian.net/wiki/spaces/SD/pages/1250951242/ship

[OSDEV-2719]: https://opensupplyhub.atlassian.net/browse/OSDEV-2719?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by TylerHeath1** [comment]
@daniellekain — opening this for review when you have a moment. This is the third Common Dev Brain PR (after #1011 explain-code and #1012 plan-ticket).

Same pattern as the others:
- Canonical skill body at `.agent/skills/ship/SKILL.md` (works for Cursor and Claude)
- Thin command wrapper at `.claude/commands/ship.md`

A few design choices worth flagging:
- /ship orchestrates the existing `oshub-codebase-expert`, `pr-description`, and `release-notes` skills rather than restating their conventions. Single source of truth.
- 8 user checkpoints throughout — never proceeds past plan, files review, tests, push, preview, or PR creation without confirmation. Designed to be safe at every experience level.
- Language calibrates to user experience: Level 1 (never coded) gets plain-English glosses for git/commit/push/PR/merge; Level 3 (engineer) gets full technical depth.
- Karpathy principles baked in (surface assumptions, minimum work, surgical changes, verifiable checks at each step) without quoting the text.

Three follow-ups queued for after this lands (won't bloat this PR):
1. Add `memory: project` to the Codebase Expert sub-agent
2. Worktree isolation for parallel implementation agents
3. PostToolUse lint hook on implementation agents

Confluence page synced to match: https://opensupplyhub.atlassian.net/wiki/spaces/SD/pages/1250951242/ship

File: .agent/skills/ship/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+---
+
+## Step 11 — Push to GitHub (CHECKPOINT)
+
+*Gloss for Levels 1/2:* "**Pushing** uploads your local commits to GitHub so others *could* see them, but doesn't ask anyone to review yet."
+
+> Ready to upload your branch to GitHub. This makes it visible online but does NOT create a PR yet. Push? (yes / no)
+
+Wait. Then `git push -u origin HEAD`.
+
```
[the remark this answers is not in the document]
Comment (daniellekain) [reply .agent/skills/ship/SKILL.md:208]: this could help make the command more explicit but I would say it's optional

File: .agent/skills/ship/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+**Generate the PR description by calling the `pr-description` skill** (`.agent/skills/pr-description/SKILL.md`). It produces the body following OS Hub conventions (Summary / How to test / Jira link, correct formatting). Do not write the description from scratch — and do not inline the convention rules here, they live in the skill.
+
+Show:
+> PR title: [OSDEV-####] [summary]
+> PR description (first 2 lines): [...]
+> Files changed: [list]
+> Create this PR? (yes / no)
+
+Wait. Then `gh pr create` with the title and the description from `pr-description`.
+
```
[the remark this answers is not in the document]
Comment (daniellekain) [reply .agent/skills/ship/SKILL.md:251]: this seems similar to the first comment, could go either way

**Review by daniellekain — DISMISSED** [review]

**Review by protsack-stephan — DISMISSED** [review]

File: .agent/skills/ship/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+- Every checkpoint has a verifiable check — name it, then wait.
+
+---
+
+## Step 0 — Route the entry
+
+Read `$ARGUMENTS`. Decide one of three paths:
+
+- **Empty** → ask: *"Which ticket are we shipping? Give me an OSDEV key (e.g., OSDEV-2316), or describe the idea and I'll help you draft a ticket first."*
+- **Matches `OSDEV-####`** → continue to Step 2.
```
Comment (jamesjohnson4395) [inline .agent/skills/ship/SKILL.md]: Should be step 1

File: .agent/skills/ship/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+## Step 1 — Confirm experience level (one-time per session)
+
+If unknown from earlier in this session, ask:
+
+> Quick question so I pitch this right:
+> 1. First time using Claude Code at OS Hub — explain every term, slow pacing
+> 2. Used it a few times — some comfort with git and the workflow
+> 3. Engineer — full technical depth, skip the hand-holding
+>
+> Which fits? (1, 2, or 3)
```
Comment (jamesjohnson4395) [inline .agent/skills/ship/SKILL.md]: This question isn't super intuitive. Consider changing to 'Which fits **your experience level?**'

File: .agent/skills/ship/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+Calibrate the run:
+- **Level 1:** Gloss every git/GitHub term at first appearance. Plain English before every shell command. Slow pacing, confidence-building.
+- **Level 2:** Gloss only non-obvious terms (PR, merge). Explain only non-routine commands.
```
Comment (jamesjohnson4395) [inline .agent/skills/ship/SKILL.md]: Are PR and merge here examples of obvious or non-obvious commands?

File: .agent/skills/ship/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+> Cached plan found at ~/.claude/plans/$ARGUMENTS.md.
+> - Created: [date]
+> - Complexity: [RED / YELLOW / GREEN]
+> - Files to change: [count]
+> - Tests to write: [count]
+>
+> Use this plan? Reply 'yes' to skip planning and jump to building, or 'no' to re-plan from scratch.
+
+If yes: skip Steps 3–6 and jump to Step 7.
```
Comment (jamesjohnson4395) [inline .agent/skills/ship/SKILL.md:71]: Instead of Steps 3-6, could we just use the /plan skill? That way, if the way we plan changes, we don't need to update it in both /plan and here.

File: .agent/skills/ship/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+## Step 7 — Create a fresh branch and implement
+
+*Confidence frame for Levels 1/2:* "I'm pulling the latest version of the team's code so you start from a clean copy. Then I'll make a **branch** — your own copy where you can make changes safely."
+
+Automatically:
+- Pull main fresh (no stale checkouts)
+- Create branch `OSDEV-####-short-description` (derive from the real ticket title; no placeholder text)
+- Verify the new branch is checked out before continuing
+
+Implement. If both backend and frontend are needed, launch two generic Agents in parallel (run_in_background: true). Each gets: the approved plan, Confluence context, recent-PR patterns, and AGENTS.md conventions. Each is told: **touch only what the plan calls for. No refactoring of adjacent code. No formatting changes to files outside the plan. Never run git commands.**
```
Comment (jamesjohnson4395) [inline .agent/skills/ship/SKILL.md]: Is there evidence that this is the best way to do development? The components can still touch each other at points, and I worry that having two separate agents running at the same time might result in code being built in disjointed ways, even with a common plan.

File: .agent/skills/ship/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+Wait for the user.
+
+---
+
+## Step 9 — Write tests, run tests, lint
+
+**Future migration:** When the `test-writer` skill exists, this step calls into it. For now, the orchestrator inlines the test-writing logic.
+
+Write tests:
+- For each new or modified Django view/model/serializer: write a happy-path test plus one edge case. Place in `src/django/api/tests/test_<feature>.py` following recent PR patterns.
```
Comment (jamesjohnson4395) [inline .agent/skills/ship/SKILL.md]: Is one edge case common practice here? If there are multiple edge cases for a feature, we should try to test all of them, no?

File: .agent/skills/ship/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+---
+
+## Step 10 — Commit locally and update release notes
+
+*Gloss for Levels 1/2:* "A **commit** is a saved snapshot of your changes on your computer. Nothing is visible to anyone else yet."
+
+- `git add` the changed files
+- `git commit -m "[OSDEV-####] [real ticket summary]"`
+- **Call the `release-notes` skill** (`.agent/skills/release-notes/SKILL.md`) to update `doc/release/RELEASE-NOTES.md` following OS Hub conventions. Do not write the entry by hand.
+- `git commit -m "[OSDEV-####] Update release notes"`
```
Comment (jamesjohnson4395) [inline .agent/skills/ship/SKILL.md:197]: I believe there would need to be another git add here, unless this exists in the release-notes skill already. The agent could probably infer this on its own, but it might not hurt to include if the agent starts wanting to follow these instructions very exactly.

File: .agent/skills/ship/SKILL.md [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+Fix pre-flight failures silently. Note any temporary changes to revert before the next commit.
+
+**Never ask the user to use browser DevTools, the console, or the Redux inspector.** Get debug info programmatically.
+
+If local preview cannot run for a structural reason (seed data mismatch, missing test account that can't be created, etc.), say so honestly. Don't claim "looks good" when you couldn't actually verify. Let the user decide whether to skip preview or address the blocker.
+
+Show:
+> Preview your change: http://localhost:6543/[path]
+> Log in as: [email] / [password]
+> What you should see: [plain English description]
```
[the remark this answers is not in the document]
Comment (jamesjohnson4395) [reply .agent/skills/ship/SKILL.md:229]: Meh, I don't _think_ the test credentials are usable on our platform anyways

File: .agent/skills/ship/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+> How to verify: [1-2 simple visual checks]
```
Comment (jamesjohnson4395) [inline .agent/skills/ship/SKILL.md]: This could be more than 1-2 for some features, idk if we want to put this limit here.

**Review by jamesjohnson4395 — CHANGES_REQUESTED** [review]

**Comment by TylerHeath1** [comment]
Thanks for the thorough review, James — really useful, and you caught two real bugs. Here's how I addressed each comment (pushed in the latest commit; the Confluence mirror is synced too):

**Fixed in code:**
- **Step 0 routing (line 26)** — good catch, this was a real bug. A valid `OSDEV-####` jumped straight to Step 2, skipping the experience-level check (Step 1). Now routes to Step 1.
- **Release-notes commit (line 192)** — also a real bug. Added `git add doc/release/RELEASE-NOTES.md` before the commit, so the release-notes change actually gets staged. Nice catch.
- **Experience-level prompt (line 40)** — reworded to "Which fits your experience level? (1, 2, or 3)".
- **Level 2 glossing (line 44)** — clarified: "Gloss only the less-familiar terms (e.g., PR, merge); skip the ones they clearly know."
- **Edge cases (line 159)** — changed from "one edge case" to "the meaningful edge cases (don't stop at one if the feature has several)."
- **Visual checks (line 225)** — softened "1-2" to "a few simple visual checks — cover what the change actually affects."

**Parallel agents (line 136)** — you were right to push on this. The real risk isn't file collisions (backend/frontend live in different folders) — it's the two agents drifting on the **API contract**. So instead of dropping parallelism, I made it **contract-first**: the plan has to pin the exact interface (endpoint, request/response shape, field names) *before* the agents split, both build to that fixed contract, and there's an integration check after they finish to confirm the frontend's calls match the backend's actual implementation. That keeps the parallel speed while closing the drift gap you flagged.

**Deferred (logged in OSDEV-2871):**
- **Steps 3–6 vs `/plan` (line 71)** — agreed, this is the same modularity point you raised on the other PRs. Rather than duplicate the planning logic, the fix is to have `/ship` call `/plan`. Captured in **OSDEV-2871** along with the other shared-pipeline-step extraction, so we can do it in one pass: https://opensupplyhub.atlassian.net/browse/OSDEV-2871

**No change:**
- **Test credentials (line 224)** — agreed they're likely not usable on the platform anyway; left as-is since it's just a placeholder in the preview step.

Re-requesting your review when you have a sec — thanks again.

**Review by jamesjohnson4395 — APPROVED** [review]

**Review by daniellekain — APPROVED** [review]
thanks for creating and addressing feedback!
